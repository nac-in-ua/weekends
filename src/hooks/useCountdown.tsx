import { useState, useEffect } from 'react'
import { getTimeLeft } from '../utils/getTimeLeft'
import { ITargetDayTimeWithMinutesAnsSeconds } from '../types'

const useCountdown = (targetTime: ITargetDayTimeWithMinutesAnsSeconds) => {
  const [time, setTime] = useState(getTimeLeft(targetTime))

  useEffect(() => {
    const interval = setTimeout(() => {
      setTime(getTimeLeft(targetTime))
    }, 1000)

    return () => {
      clearInterval(interval)
    }
  }, [time, targetTime])

  return time
}

export default useCountdown
