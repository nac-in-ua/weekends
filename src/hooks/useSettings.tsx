import React from 'react'
import { SettingsDataContext, SettingsDispatchContext } from '../store/Settings'
import { ISettings, Dispatch } from '../types'

const useSettingsData = () => {
  const context = React.useContext(SettingsDataContext)
  if (context === undefined) {
    throw new Error(
      'useSettingsData must be used within a SettingsContextProvider'
    )
  }
  return context
}

const useSettingsDispatch = () => {
  const context = React.useContext(SettingsDispatchContext)
  if (context === undefined) {
    throw new Error(
      'useSettingsDispatch must be used within a SettingsContextProvider'
    )
  }
  return context
}

const useSettings = (): [ISettings, Dispatch] => {
  return [useSettingsData(), useSettingsDispatch()]
}

export { useSettings, useSettingsData, useSettingsDispatch }
