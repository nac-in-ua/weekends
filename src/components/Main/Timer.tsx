import Clock from '../Clock/Clock'
import Actiontext from './Actiontext'
import { useEffect } from 'react'
import useSettings from '../../hooks/useSettings'
import { getTimerData } from '../../utils/settings'
import { isFinished } from '../../utils/getTimeLeft'
import useCountdown from '../../hooks/useCountdown'

type TimerProps = {
  onFinish: Function
}

function Timer({ onFinish }: TimerProps) {
  const [settings] = useSettings()
  const time = useCountdown(getTimerData(settings))

  useEffect(() => {
    if (isFinished(time)) {
      onFinish(true)
    }
  }, [time, onFinish])

  return (
    <div className="mt-20 flex flex-col items-center">
      <Actiontext />
      <Clock time={time} />
    </div>
  )
}

export default Timer
