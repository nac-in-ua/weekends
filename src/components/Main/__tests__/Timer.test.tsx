import { render, screen } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import Timer from '../Timer'
import { useSettings } from '../../../hooks/useSettings'

jest.mock('../../../hooks/useSettings')

describe('Timer', () => {
  const handleDispatch = jest.fn()
  const onFinishHandler = jest.fn()
  const mockedUseSettings = jest.mocked(useSettings)

  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.useRealTimers()
    onFinishHandler.mockClear()
    mockedUseSettings.mockClear()
  })

  it('should render', () => {
    const date = new Date(2022, 4, 23, 18, 12, 11, 0)
    jest.setSystemTime(date)
    mockedUseSettings.mockImplementation(() => [
      {
        day: 5,
        hour: 18,
        greetingsText: 'sample',
        theme: 'light',
        useSystemTheme: false,
        isFirstLoad: false,
      },
      handleDispatch,
    ])
    const { container } = render(<Timer onFinish={onFinishHandler} />)
    expect(screen.getByText(/Time left to weekends/i)).toBeInTheDocument()
    expect(screen.getByText(/seconds/i)).toBeInTheDocument()
    expect(onFinishHandler).not.toBeCalled()
    expect(container).toMatchSnapshot()
  })

  it('should trigger onFinish when time finishes', () => {
    const date = new Date(2022, 4, 23, 18, 0, 0, 0)
    jest.setSystemTime(date)
    mockedUseSettings.mockImplementation(() => [
      {
        day: 1,
        hour: 18,
        greetingsText: 'sample',
        theme: 'light',
        useSystemTheme: false,
        isFirstLoad: false,
      },
      handleDispatch,
    ])
    render(<Timer onFinish={onFinishHandler} />)
    expect(onFinishHandler).toBeCalled()
  })

  it('should trigger onFinish on next tick', () => {
    const date = new Date(2022, 4, 23, 17, 59, 59, 0)
    jest.setSystemTime(date)
    mockedUseSettings.mockImplementation(() => [
      {
        day: 1,
        hour: 18,
        greetingsText: 'sample',
        theme: 'light',
        useSystemTheme: false,
        isFirstLoad: false,
      },
      handleDispatch,
    ])
    render(<Timer onFinish={onFinishHandler} />)
    const secondsLeft = screen.getByTestId(/digit/i)
    expect(secondsLeft).toHaveTextContent('1')
    act(() => {
      jest.runOnlyPendingTimers()
    })
    expect(secondsLeft).toHaveTextContent('0')
    expect(onFinishHandler).toHaveBeenCalledTimes(1)
  })
})
