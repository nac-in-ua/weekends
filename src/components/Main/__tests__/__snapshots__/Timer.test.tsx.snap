// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer should render 1`] = `
<div>
  <div
    class="mt-20 flex flex-col items-center"
  >
    <h2
      class="text-center text-3xl font-thin uppercase"
    >
      Time left to weekends
    </h2>
    <div
      class="mt-10 flex flex-row justify-around"
      data-cy="clock"
    >
      <div
        class="flex w-24 flex-col"
      >
        <div
          class="flex h-8 items-center justify-center text-lg font-thin uppercase"
        >
          days
        </div>
        <div
          class="flex h-20 items-center justify-center text-5xl font-thin"
          data-testid="digit"
        >
          3
        </div>
      </div>
      <div
        class="mt-8 flex items-center justify-center pb-2 text-5xl"
      >
        :
      </div>
      <div
        class="flex w-24 flex-col"
      >
        <div
          class="flex h-8 items-center justify-center text-lg font-thin uppercase"
        >
          hours
        </div>
        <div
          class="flex h-20 items-center justify-center text-5xl font-thin"
          data-testid="digit"
        >
          23
        </div>
      </div>
      <div
        class="mt-8 flex items-center justify-center pb-2 text-5xl"
      >
        :
      </div>
      <div
        class="flex w-24 flex-col"
      >
        <div
          class="flex h-8 items-center justify-center text-lg font-thin uppercase"
        >
          minutes
        </div>
        <div
          class="flex h-20 items-center justify-center text-5xl font-thin"
          data-testid="digit"
        >
          47
        </div>
      </div>
      <div
        class="mt-8 flex items-center justify-center pb-2 text-5xl"
      >
        :
      </div>
      <div
        class="flex w-24 flex-col"
      >
        <div
          class="flex h-8 items-center justify-center text-lg font-thin uppercase"
        >
          seconds
        </div>
        <div
          class="flex h-20 items-center justify-center text-5xl font-thin"
          data-testid="digit"
        >
          49
        </div>
      </div>
    </div>
  </div>
</div>
`;
